ipa-core/src/protocol/ipa_prf/oprf_padding/mod.rs

```diff
@@ -1,5 +1,814 @@
+        };
+        assert_eq!(from_right, from_left);
+        total_number_of_fake_rows = from_right;
+    }
+
+    // Step 3: `h_out` will generate secret shares of zero for as many rows as the `total_number_of_fake_rows`
+    if ctx.role() == h_out {
+        for _ in 0..total_number_of_fake_rows as usize {
+            let row = OPRFIPAInputRow {
+                match_key: Replicated::new(BA64::ZERO, BA64::ZERO),
+                is_trigger: Replicated::new(Boolean::FALSE, Boolean::FALSE),
+                breakdown_key: Replicated::new(BK::ZERO, BK::ZERO),
+                trigger_value: Replicated::new(TV::ZERO, TV::ZERO),
+                timestamp: Replicated::new(TS::ZERO, TS::ZERO),
+            };
+            padding_input_rows.push(row);
+        }
+    }
+
+    input.extend(padding_input_rows);
+    Ok(input)
+}
+
+/// # Errors
+/// Will propogate errors from `OPRFPaddingDp`
+/// # Panics
+///
+pub fn two_parties_add_dummies<C, BK, TV, TS, const B: usize>(
+    ctx: &C,
+    padding_input_rows: &mut Vec<OPRFIPAInputRow<BK, TV, TS>>,
+    h_i: Role,
+    h_i_plus_one: Role,
+    padding_params: &PaddingParameters,
+) -> Result<u32, Error>
+where
+    C: Context,
+    BK: BooleanArray + U128Conversions,
+    TV: BooleanArray,
+    TS: BooleanArray,
+{
+    let mut total_number_of_fake_rows = 0;
+    let (mut left, mut right) = ctx.prss_rng();
+    // The first is shared with the helper to the "left", the second is shared with the helper to the "right".
+    let mut rng = &mut right;
+    if ctx.role() == h_i {
+        rng = &mut right;
+    }
+    if ctx.role() == h_i_plus_one {
+        rng = &mut left;
+    }
+
+    // padding for oprf
+    match padding_params.oprf_padding {
+        OPRFPadding::NoOPRFPadding => {}
+        OPRFPadding::Parameters {
+            oprf_epsilon,
+            oprf_delta,
+            matchkey_cardinality_cap,
+            oprf_padding_sensitivity,
+        } => {
+            let oprf_padding =
+                OPRFPaddingDp::new(oprf_epsilon, oprf_delta, oprf_padding_sensitivity)?;
+            for cardinality in 1..=matchkey_cardinality_cap {
+                let sample = oprf_padding.sample(rng);
+                total_number_of_fake_rows += sample * cardinality;
+
+                // this means there will be `sample` many unique
+                // matchkeys to add each with cardinality = `cardinality`
+                for _ in 0..sample {
+                    let dummy_mk: BA64 = rng.gen();
+                    for _ in 0..cardinality {
+                        let mut match_key_shares: Replicated<BA64> = Replicated::default();
+                        if ctx.role() == h_i {
+                            match_key_shares = Replicated::new(BA64::ZERO, dummy_mk);
+                        }
+                        if ctx.role() == h_i_plus_one {
+                            match_key_shares = Replicated::new(dummy_mk, BA64::ZERO);
+                        }
```

Comment:
Again, I think you can potentially avoid mutability with a match expression.

Reply:
with h_i and h_i_plus_one variables, it doesn't seem to want to match as follows, saying the first one might "match any value"

```
                        let match_key_shares: Replicated<BA64> = match ctx.role() {
                            _h_i => Replicated::new(BA64::ZERO, dummy_mk),
                            _h_i_plus_one => Replicated::new(dummy_mk, BA64::ZERO),
                            _ => Replicated::default(),
                        };

                        let row = OPRFIPAInputRow {
                            match_key: match_key_shares,
                            is_trigger: Replicated::new(Boolean::FALSE, Boolean::FALSE),
                            breakdown_key: Replicated::new(BK::ZERO, BK::ZERO),
                            trigger_value: Replicated::new(TV::ZERO, TV::ZERO),
                            timestamp: Replicated::new(TS::ZERO, TS::ZERO),
                        };
                        
error: unreachable pattern
   --> ipa-core/src/protocol/ipa_prf/oprf_padding/mod.rs:349:29
    |
348 | ...                   _h_i => Replicated::new(BA64::ZERO, dummy_mk),
    |                       ---- matches any value
349 | ...                   _h_i_plus_one => Replicated::new(dummy_mk, BA64::ZERO),
    |                       ^^^^^^^^^^^^^ unreachable pattern